### DISCUSSION TITLE
How to use ErrorFormatterWithPageUrl

### DISCUSSION THREAD

**Discussion by navhaxs** [selenide/selenide#2830] [Q&A] [answered]
I want the Page URL included whenever a Selenide test fails.

I found this "ErrorFormatterWithPageUrl" class in https://github.com/selenide/selenide/blob/main/modules/error-message-customizer/src/main/java/com/codeborne/selenide/ex/custom/ErrorFormatterWithPageUrl.java

Wondering how to use this (example?) code? Is this something I should copy into my own project? Do I need to configure something so that Selenide will use set this class as the error formatter?

Thanks :)

**Comment by navhaxs** [comment, marked as answer]
Ah it looks like this is part of the plugin system. So I have to create a file `resources/META-INF/services/com.codeborne.selenide.ex.ErrorFormatter` with the classname as the content!

**Reply by navhaxs** [reply]
Wanted to add, whilst this method is working nicely for customising errors printed to the console, the `SelenideErrorFormatter` has an issue when attempting to write to Gradle's test report HTML.

It seems that this happens on a different thread and so calling `driver.url()` fails since it doesn't have access to the Driver object :(

e.g.
```
java.lang.IllegalStateException: No webdriver is bound to current thread: 19. You need to call open(url) first.
```

So whilst customising the error message works for the console output, it doesn't for the Gradle test report html.

**Reply by BorisOsipov** [reply]
>SelenideErrorFormatter has an issue when attempting to write to Gradle's test report HTML.
>So whilst customising the error message works for the console output, it doesn't for the Gradle test report html.

SelenideErrorFormatter has no idea about gradle, html, console and so on. It simply adds the URL string to the error stacktrace message... How you use stacktrace it is your thing..

**Reply by navhaxs** [reply]
hmm yeah thought so. I'll see what I can do on my end.
Thanks for the recent v7.4.3 release btw :)

**Reply by asolntsev** [reply]
@BorisOsipov Well, I guess it's still a problem on Selenide side. 
I see that `SelenideErrorFormatter` is called only from method `getMessage()`, and this can be called after closing webdriver. 
I think we need to fix it so that `SelenideErrorFormatter.generateErrorDetails` is called when throwing the exception.

**Reply by BorisOsipov** [reply]
@asolntsev 
>I see that SelenideErrorFormatter is called only from method getMessage(), and this can be called after closing webdriver.

Sounds like a really bad thing.

**Reply by asolntsev** [reply]
Will be fixed by https://github.com/selenide/selenide/pull/2839
